### PR TITLE
ISPN-11875 SQL Store

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-11875

### CODE REVIEW

**Comment by wburns** [comment]
This should be reviewable now. CI should be okay, however there are a few more tests I want to add to close some more edge cases.

**Comment by wburns** [comment]
One thing I noticed while writing tests is it seems we may want a package for key and value separately instead of assuming the package is the same for both the key and value at the same time.

File: persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/UnitTestDatabaseManager.java [hunk 1]
```diff
@@ -28,14 +30,14 @@
    private static final String DB_TYPE = System.getProperty("infinispan.test.jdbc.db", "H2");
    private static final String H2_DRIVER = org.h2.Driver.class.getName();
    private static final String NON_EXISTENT_DRIVER = "non.existent.Driver";
-   private static final DatabaseType dt;
+   private static final org.infinispan.persistence.jdbc.common.DatabaseType dt;
```
Comment (ryanemerson) [inline persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/UnitTestDatabaseManager.java:33]: No need for the FQN.

File: persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStoreTest.java [hunk 2]
```diff
@@ -50,61 +37,24 @@ protected String parameters() {
...
-      tableManager.stop();
-
-      stringBasedCacheStore.stop();
+      return configurationBuilder.build();
    }
 
    @Override
    protected boolean storePurgesAllExpired() {
-      // expiration listener is not called for the entries
+      // TODO: we need to add back in expiration support
```
Comment (ryanemerson) [inline persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStoreTest.java:57]: Don't forget :slightly_smiling_face:

File: persistence/remote/src/test/java/org/infinispan/persistence/remote/RemoteStoreTest.java [hunk 3]
```diff
@@ -146,8 +146,9 @@ protected PersistenceMarshaller getMarshaller() {
    }
 
    @Override
-   @AfterMethod
+   @AfterMethod(alwaysRun = true)
    public void tearDown() {
+      configuration = null;
```
Comment (ryanemerson) [inline persistence/remote/src/test/java/org/infinispan/persistence/remote/RemoteStoreTest.java:151]: Why are the changes required for the `persistence/remote` tests?

**Review by ryanemerson — COMMENTED** [review]
@wburns Looking good so far! I have reviewed the jdbc-common and jdbc changes. I'll take a look at the new store shortly. 

Don't forget to implement the various TODOs on the StoreMigrator :slightly_smiling_face:

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+
+      assert Arrays.stream(parameters).filter(Parameter::isPrimaryIdentifier).noneMatch(parameter -> parameter.unwrapJsonKey == null);
+   }
+
+   Parameter[] handleUnusedValueParams(Parameter[] parameters, List<Parameter> unusedValueParams) {
+      // If it is a loader, we can ignore missing values as it is read only
+      if (!config.ignoreModifications()) {
+         throw unusedValueParamsException(unusedValueParams);
+      }
+      // TODO: add debug message here to say that columns are missing from schema object
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:199]: Shouldn't we throw an exception here? If columns are missing from the schema this is probably a mistake from the user .. or is this intentional so that users can avoid specifying all fields in tables with many columns?

File: build-configuration/bom/pom.xml [hunk 5]
```diff
@@ -91,6 +91,22 @@
...
+            <dependency>
+                <groupId>org.infinispan</groupId>
+                <artifactId>infinispan-cachestore-jdbc-common</artifactId>
+                <version>${version.infinispan}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.infinispan</groupId>
+                <artifactId>infinispan-cachestore-jdbc-common</artifactId>
+                <version>${version.infinispan}</version>
+                <type>test-jar</type>
```
[the remark this answers is not in the document]
Comment (wburns) [reply build-configuration/bom/pom.xml:103]: I would prefer to do this, but then I have to define the version somewhere else. Do you know where we normally do that?

File: core/src/main/java/org/infinispan/marshall/persistence/PersistenceMarshaller.java [hunk 6]
```diff
@@ -23,6 +24,8 @@
     */
    void register(SerializationContextInitializer initializer);
 
+   ImmutableSerializationContext getSerializationContext();
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/marshall/persistence/PersistenceMarshaller.java:27]: It is only unused because I had to do the hacky thing to get the user context. If this is the way I should be doing it, then I am fine removing it. https://github.com/infinispan/infinispan/pull/9480/files#diff-05c73e851df1a197533a5e98b9d28ed7a95b31ead0ff6dadc09e4d23e563fd23R77

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+
+   private void updateUnwrap(Parameter parameter, boolean key, Function<Json, Json> function) {
+      if (key) {
+         parameter.unwrapJsonKey = function;
+      } else {
+         parameter.unwrapJsonValue = function;
+      }
+   }
+
+   protected enum SchemaType {
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:331]: WDYT to changing this to `SchemaFieldType` or just `FieldType`?

File: core/src/test/java/org/infinispan/persistence/BaseStoreFunctionalTest.java [hunk 8]
```diff
@@ -149,36 +156,65 @@ public void testPreloadAndExpiry() {
...
+      Person person2 = createEmptyPerson("2");
+      person2.setPicture(pictureBytes);
+      assertEquals(person2, cache.get("k2"));
+      Person person3 = createEmptyPerson("3");
+      person3.setSex(Sex.MALE);
+      assertEquals(person3, cache.get("k3"));
+      assertEquals(new Person("4", new Address("Street", "City", 12345), null, null, null), cache.get("k4"));
+      Person infinispanPerson = createEmptyPerson("Infinispan");
+      infinispanPerson.setBirthDate(infinispanBirthDate);
+      assertEquals(infinispanPerson, cache.get("Infinispan"));
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/BaseStoreFunctionalTest.java:194]: For this test, I needed it slightly different as the protoschema has some required fields (Date) so I couldn't use just the constructor. And other stores do not rely upon that so it didn't require those fields. Check the overridden method
`return new Person(name, new Address(), null, null, null);` vs `return new Person(name);`

File: core/src/test/java/org/infinispan/persistence/SegmentedStoreTest.java [hunk 9]
```diff
@@ -1,112 +0,0 @@
...
-
-import io.reactivex.rxjava3.core.Flowable;
-
-/**
- * Test that ensures that various segment based methods work properly
- * @author William Burns
- * @since 9.4
- */
-@Test (groups = "functional", testName = "persistence.SegmentedStoreTest")
-public abstract class SegmentedStoreTest extends SingleCacheManagerTest {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/SegmentedStoreTest.java:32]: This was only used by `JdbcStringBasedSegmentedStoreTest` which uses the old SPI. And with the new SPI it doesn't really make any sense anymore as all these methods have been condensed. We already test segmentation for the various methods in `JdbcStringBasedStoreTest` which extends `BaseNonBlockingStoreTest` that should be the place to add new methods as needed.

File: core/src/test/java/org/infinispan/test/data/Key.java [hunk 10]
```diff
@@ -12,32 +12,35 @@
    private static final long serialVersionUID = 4745232904453872125L;
 
    @ProtoField(number = 1)
-   String value;
+   String keyColumn;
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/test/data/Key.java:15]: Yeah, the main issue is my database was quite weird.  I can change the tests a bit to tweak it back.

File: persistence/jdbc-common/src/test/java/org/infinispan/persistence/jdbc/common/UnitTestDatabaseManager.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * Class that assures concurrent access to the in memory database.
+ *
+ * @author Mircea.Markus@jboss.com
+ * @author Navin Surtani (<a href="mailto:nsurtani@redhat.com">nsurtani@redhat.com</a>)
+ * @author Tristan Tarrant
+ */
+
+public class UnitTestDatabaseManager {
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc-common/src/test/java/org/infinispan/persistence/jdbc/common/UnitTestDatabaseManager.java:21]: Yes, there is a method in the jdbc one that couldn't go into the commons module. I can remove the overlap though.

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/DatabaseStatementsManager.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import org.infinispan.persistence.jdbc.common.impl.table.GenericDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.H2DatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.MySQLDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.OracleDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.PostgresqlDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.SQLLiteDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.SQLServerDatabaseStatementsManager;
+import org.infinispan.persistence.jdbc.common.impl.table.SybaseDatabaseStatementsManager;
+
+public interface DatabaseStatementsManager {
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/DatabaseStatementsManager.java:15]: I am totally fine with changing the names. I had to evolve some things over time and didn't update the old names. There wasn't this statements manager originally, until I found I really needed to abstract the statements solely for the different DB providers.

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 13]
```diff
@@ -30,7 +44,7 @@
 /**
  * @author Ryan Emerson
  */
-public abstract class AbstractTableManager implements TableManager {
+public abstract class AbstractTableManager<K, V> extends BaseSqlManager<K, V> implements TableManager<K, V> {
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:47]: I wanted to, but it is a bit more work. I was thinking of removing them in a subsequent PR, I figured that would cause less churn as well.

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/AbstractJdbcStoreConfiguration.java [hunk 14]
```diff
@@ -43,8 +40,12 @@ public ConnectionFactoryConfiguration connectionFactory() {
       return connectionFactory;
    }
 
+   /**
+    * @return always returns false
+    * @deprecated since 13.0
+    */
    public boolean manageConnectionFactory() {
-      return manageConnectionFactory.get();
+      return false;
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/AbstractJdbcStoreConfiguration.java:48]: No, but this was never a feature that was available. It was never part of the xml configuration. But I can make one to link to it.

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/Element.java [hunk 15]
```diff
@@ -27,6 +29,17 @@
...
+   SCHEMA("schema"),
+
+   QUERIES("queries"),
+   SELECT_SINGLE("select-single"),
+   SELECT_ALL("select-all"),
+   DELETE_SINGLE("delete-single"),
+   DELETE_ALL("delete-all"),
+   UPSERT("upsert"),
+   SIZE("size"),
+
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/Element.java:42]: That was my original thought that it was a bit overkill to have 2 more element classes (as common will still want one) and we could reevaluate this later if needed.

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/impl/BaseJdbcStore.java [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+import org.infinispan.persistence.spi.NonBlockingStore;
+import org.infinispan.persistence.spi.PersistenceException;
+import org.infinispan.util.concurrent.BlockingManager;
+import org.infinispan.util.logging.LogFactory;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.internal.functions.Functions;
+
+public abstract class BaseJdbcStore<K, V, C extends AbstractJdbcStoreConfiguration> implements NonBlockingStore<K, V> {
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/impl/BaseJdbcStore.java:32]: It doesn't, the generics are a bit confusing on the configuration classes :D

File: persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/UnitTestDatabaseManager.java (hunk 1 above)
Comment (wburns) [reply persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/UnitTestDatabaseManager.java:33]: This was removed with the refactor of the two.

File: persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStoreTest.java (hunk 2 above)
Comment (wburns) [reply persistence/jdbc/src/test/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStoreTest.java:57]: I actually implemented it! I however did not change this test as the new expiration code works properly to expose the entry unlike the old expiration code. Let me see if just removing this method works, heh.

File: persistence/remote/src/test/java/org/infinispan/persistence/remote/RemoteStoreTest.java (hunk 3 above)
Comment (wburns) [reply persistence/remote/src/test/java/org/infinispan/persistence/remote/RemoteStoreTest.java:151]: I changed the base store to not create a config every time, this was required for some of the database tests as each time we created a new config it would create a unique connection url. By setting the configuration to null ensures this test creates  a new config after each test, which is required for it to pass.

File: persistence/sql/pom.xml [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+   <dependencies>
+      <dependency>
+         <groupId>org.infinispan.protostream</groupId>
+         <artifactId>protostream-processor</artifactId>
+      </dependency>
+
+      <dependency>
+         <groupId>com.fasterxml.jackson.core</groupId>
+         <artifactId>jackson-core</artifactId>
+      </dependency>
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/pom.xml:24]: It is sadly to do any conversions between json and other types as was discussed in https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/sql.20store/near/247707385 because of https://github.com/infinispan/protostream/blob/main/core/src/main/java/org/infinispan/protostream/impl/JsonUtils.java#L51

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+
+      parameters = generateParameterInformation(config, connectionFactory);
+      assert parameters.length != 0;
+      primaryParameters = determinePrimaryParameters(config, parameters);
+      assert primaryParameters.length != 0;
+      assert Arrays.stream(primaryParameters).allMatch(Parameter::isPrimaryIdentifier);
+
+      persistenceMarshaller = ctx.getPersistenceMarshaller();
+      // TODO: we need to figure out what to do with the context here
+//      serializationContext = persistenceMarshaller.getSerializationContext();
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:77]: This may or may not be needed, depending upon how we should do the serialization context stuff. I am guessing I still need to remove this and instead use the user context and if not then the system context, correct?

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/TableJdbcStoreConfigurationBuilder.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+   @Override
+   public TableJdbcStoreConfigurationBuilder self() {
+      return this;
+   }
+
+   @Override
+   public void validate(GlobalConfiguration globalConfig) {
+      super.validate(globalConfig);
+      if (attributes.attribute(TableJdbcStoreConfiguration.TABLE_NAME).get() == null) {
+         // TODO: make a
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/TableJdbcStoreConfigurationBuilder.java:34]: ?

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/SqlSerializerUtil.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
+package org.infinispan.persistence.sql.configuration;
+
+import org.infinispan.commons.configuration.io.ConfigurationWriter;
+import org.infinispan.persistence.jdbc.common.configuration.Element;
+
+public class SqlSerializerUtil {
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/SqlSerializerUtil.java:6]: package private?

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/SchemaJdbcConfigurationBuilder.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+public class SchemaJdbcConfigurationBuilder implements Builder<SchemaJdbcConfiguration> {
+
+   private final AttributeSet attributes;
+
+   public SchemaJdbcConfigurationBuilder() {
+      this.attributes = SchemaJdbcConfiguration.attributeDefinitionSet();
+   }
+
+   AttributeSet attributes() {
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/SchemaJdbcConfigurationBuilder.java:22]: Never used.

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/AbstactSchemaJdbcConfigurationBuilder.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import org.infinispan.commons.CacheConfigurationException;
+import org.infinispan.commons.configuration.Builder;
+import org.infinispan.commons.configuration.attributes.Attribute;
+import org.infinispan.commons.configuration.attributes.AttributeSet;
+import org.infinispan.commons.logging.Log;
+import org.infinispan.configuration.cache.AbstractStoreConfiguration;
+import org.infinispan.configuration.cache.PersistenceConfigurationBuilder;
+import org.infinispan.persistence.jdbc.common.configuration.AbstractJdbcStoreConfigurationBuilder;
+
+public abstract class AbstactSchemaJdbcConfigurationBuilder<T extends AbstractSchemaJdbcConfiguration,
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/configuration/AbstactSchemaJdbcConfigurationBuilder.java:12]: `AbstractSchema...`

File: persistence/sql/src/test/java/org/infinispan/persistence/sql/AbstractSQLStoreFunctionalTest.java [hunk 23]
```diff
@@ -0,0 +1,449 @@
...
+   }
+
+   public void testDBHasMoreKeyColumnsWithNoKeySchema(Method m) {
+      Exceptions.expectException(CacheConfigurationException.class, CompletionException.class,
+            CacheConfigurationException.class,
+            "Primary key has multiple columns .*",
+            () -> testSimpleGetAndPut(m.getName(), "key", "value"));
+   }
+
+//   public void testDBHasLessKeyColumnsWithSchema(Method m) {
```
Comment (ryanemerson) [inline persistence/sql/src/test/java/org/infinispan/persistence/sql/AbstractSQLStoreFunctionalTest.java:173]: Needs implementing still.

File: persistence/sql/src/test/java/org/infinispan/persistence/sql/AbstractSQLStoreFunctionalTest.java [hunk 24]
```diff
@@ -0,0 +1,449 @@
...
+import org.infinispan.test.data.Person;
+import org.infinispan.test.data.Sex;
+import org.infinispan.transaction.TransactionMode;
+import org.postgresql.Driver;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public abstract class AbstractSQLStoreFunctionalTest extends BaseStoreFunctionalTest {
```
Comment (ryanemerson) [inline persistence/sql/src/test/java/org/infinispan/persistence/sql/AbstractSQLStoreFunctionalTest.java:44]: It would be good if we had a similar test to this in `server/tests`. That way QE can test against different DB containers easily.

\cc @gustavolira @diegolovison

**Review by ryanemerson — COMMENTED** [review]
Don't forget the TODOs in the `*.xsd` schemas :slightly_smiling_face: .

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/QueryNamedParameterParser.java [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+         this.sqlToUse = sqlToUse;
+         this.uniqueParameters = uniqueParameters;
+         this.orderedParameters = orderedParameters;
+      }
+
+      public String getSqlToUse() {
+         return sqlToUse;
+      }
+
+      public Set<String> getUniqueParameters() {
```
Comment (ryanemerson) [inline persistence/sql/src/main/java/org/infinispan/persistence/sql/QueryNamedParameterParser.java:214]: Never used. We can remove `namedParameters` from `parseSqlStatement`

File: build-configuration/bom/pom.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply build-configuration/bom/pom.xml:103]: I think you need to define it in the root `pom.xml` `<dependencyManagement>`

File: core/src/test/java/org/infinispan/persistence/BaseStoreFunctionalTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/BaseStoreFunctionalTest.java:194]: I don't understand why the following is required?

```java
      // This is only required until we can remove the default 0
      this.birthDate = birthDate == null ? new Date(0) : birthDate;
```

You have a default value of 0 for the date field, does this not ensure that `birthDate != null` when ProtoStream creates the object via this constructor?

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/Element.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/configuration/Element.java:42]: Agreed. We can always split it out later if it's ever required (probably not).

File: persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/impl/BaseJdbcStore.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply persistence/jdbc-common/src/main/java/org/infinispan/persistence/jdbc/common/impl/BaseJdbcStore.java:32]: :man_facepalming: 

I completely misread the signature :sweat_smile:

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryanemerson) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:77]: > I am guessing I still need to remove this and instead use the user context and if not then the system context, correct?

Thinking about this some more, as we're only concerned with converting user types for key/values, there shouldn't be any reason that we need to consider the persistence context here. All of our internal persistence types are known. The only thing we need to validate here is that the user's Message types exist, so checking the user ctx should suffice.

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 26]
```diff
@@ -0,0 +1,626 @@
...
+         json.set("_value", value);
+      };
+   }
+
+   void verifyParametersPresentForMessage(String fullTypeName, Map<String, Parameter> parameterMap, boolean key) {
+      GenericDescriptor genericDescriptor = null;
+      try {
+         genericDescriptor = serializationContext.getDescriptorByName(fullTypeName);
+      } catch (IllegalArgumentException t) {
+         // This is really bad for flow control TODO: add a new method to protostream
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:228]: This isn't actually as big of a deal now that I am not dynamically generating the schema as I just rethrow another exception. But I will log a JIRA to not lose track as we may need it later.

[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:228]: https://issues.redhat.com/browse/IPROTO-220

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 27]
```diff
@@ -0,0 +1,626 @@
...
+    * Method to be overridden to determine what the parameters are for the various sql statements that will be used Only
+    * the {@link #connectionFactory} will be initialized at this point
+    *
+    * @param config
+    * @return
+    * @throws SQLException
+    */
+   abstract Parameter[] generateParameterInformation(C config, ConnectionFactory connectionFactory) throws SQLException;
+
+   int typeWeUse(int sqlType, String typeName) throws SQLException {
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:106]: I was trying to keep it open for classes that may extend it, but I can remove it.

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 28]
```diff
@@ -0,0 +1,626 @@
...
+            return Types.VARBINARY;
+         }
+      }
+      return sqlType;
+   }
+
+   void verifyAndRegisterSchema(SchemaJdbcConfiguration schemaJdbcConfiguration) {
+      String packageName = schemaJdbcConfiguration.packageName();
+      String keyMessageName = schemaJdbcConfiguration.keyMessageName();
+      // We have to use a tree map to provide case insensitivity
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:119]: Yeah, I changed it to put the key as upper case instead.

File: persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java [hunk 29]
```diff
@@ -0,0 +1,626 @@
...
+      DataConversion keyDataConversion = advancedCache.getKeyDataConversion()
+            .withRequestMediaType(jsonStringType);
+      DataConversion valueDataConversion = advancedCache.getValueDataConversion()
+            .withRequestMediaType(jsonStringType);
+
+      ComponentRegistry componentRegistry = advancedCache.getComponentRegistry();
+      componentRegistry.wireDependencies(keyDataConversion, true);
+      componentRegistry.wireDependencies(valueDataConversion, true);
+
+      parameters = generateParameterInformation(config, connectionFactory);
```
[the remark this answers is not in the document]
Comment (wburns) [reply persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java:69]: Yeah, I can clean this up a bit. Went through a few different iterations :)
